pennylane/qinfo/fidelities.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Differentiable quantum fidelity"""
+
+import pennylane as qml
+
+
+def fidelity(qnode0, qnode1, wires0, wires1):
+    r"""Compute the fidelity for two for two :class:`.QNode` returning a :func:`~.state` (a state can be a state vector
+    or a density matrix, depending on the device) acting on quantum systems with the same size. For two pure states, the
+    fidelity corresponds to the squared overlap. For a pure state and a mixed state, it corresponds to the squared
+    expectation of the mixed state in the pure state. Finally for two mixed states, it is defined by the last formula:
+
+    .. math::
+        F( \ket{\psi} , \ket{\phi}) = \left|\bra{\psi}\ket{\phi}\right|^2
+
+        F( \ket{\psi} , \sigma ) = \left|\bra{\psi} \sigma \ket{\psi}\right|^2
+
+        F( \rho , \sigma ) = -\text{Tr}( \sqrt{\sqrt{\rho} \sigma \sqrt{\rho}})^2
+
+    .. warning::
+        The second state is coerced to the type and dtype of the first state. The fidelity is returned in the type
+        of the interface of the first state.
+
+    Args:
+        state0 (QNode): A :class:`.QNode` returning a :func:`~.state`.
+        state1 (QNode): A :class:`.QNode` returning a :func:`~.state`.
+        wires0 (Sequence[int]): the wires of the first subsystem
+        wires1 (Sequence[int]): the wires of the second subsystem
+
+    Returns:
+        func: A function with the same arguments as the QNode that returns the mutual information from its output state.
+
+    **Example**
```

Comment:
Maybe add an advanced example with a device of multiple qubits and you compute the overlap between parts of it. Being able to compute overlaps of reduced states is a nice feature to point out and it further explains the functionality of the `wires0` and `wires1` arguments.

Reply:
Agreed! Also, could do another device (e.g., `default.mixed`).